Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -64,21 +68,30 @@ final class ConnectionPool {
     /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
     /// is used to determine if there already exists an associated `HTTP1ConnectionProvider` in `providers`.
     /// If there is, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
-    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?, setupComplete: EventLoopFuture<Void>) -> EventLoopFuture<Connection> {
+    func getConnection(_ request: HTTPClient.Request,
+                       preference: HTTPClient.EventLoopPreference,
+                       taskEventLoop: EventLoop,
+                       deadline: NIODeadline?,
+                       setupComplete: EventLoopFuture<Void>,
+                       logger: Logger) -> EventLoopFuture<Connection> {
```

Comment:
To Konrad's point, we're losing an ability to change log level for a subsystem: imagine you have a big project, and some particular part of it misbehaves. With only one logger passed into every library, changing a log level will be all-or-nothing. Given logging is not free, we may make it impossible to enable debug logging for heavily enough loaded production systems.

Reply:
I've been meaning to ask this question as well, are we _sure_ that passing loggers around is a good idea?

Reply:
> are we sure that passing loggers around is a good idea?

I think more context is important to this question (no pun intended), as there's somewhat 2 cases I believe:

- I have some context (baggage / trace id / logger metadata) and want the lib to keep using the same as I have, e.g. a http client should "carry my trace identifiers" that this request is part of
  - ^ where there has to be _some_ form of passing around; I hope this would be addressed by the ongoing gsoc work eventually;
- "the (sub-)system" does some stuff and I want it to log things; it's not part of my "workflow" / "trace" / "request hot path", I'd want to enable "background processes logging"     
  - ^ where I'd see it as a setting

Reply:
@artemredkin / @avolokhov The question is _what is a logger?_ From SwiftLog's point of view a logger is label+metadata+logLevel+logging backend. Nothing else.

The logging backend is mostly a global constant (the configured `LoggingSystem.factory`). So really it's label+metadata+logLevel.

- We all agree (I think) that we need to pass metadata around.
- I think it's also important being able to control the log level of a sub-system that may be logging
- I think we also agree that given that we have the option of using non-standard logging backends, we need to pass that to libraries.
- It seems that many people want `label` to be controlled by the sub-system (because it's used as "source information")

If we unpack this:
- we would like to pass metadata around
- we would like to control the log-level (ie. pass it around)
- we would like to pass the logging backend around
- we would like the sub-system to control the label

The first three is pretty much everything that you achieve passing the `Logger` around, it's no different to passing the metadata and the log level around.

The sticking point seems to be `label`. In SwiftLog, when passing around a `Logger`, you also pass around the `label` and you can't change it. The label itself was added quite late in the design of SwiftLog and I never cared about it much because I mostly use one root-logger (`label: "my-app-name"`) and the go from there, passing that to everything. Adjusting metadata/logLevel as needed which works fine because `Logger` is a value type regarding metadata/log level.

The thing that is missing today is that we can't override the `label`. So I'm thinking what would be needed is either:

```
func subSystemCall(logger: Logger) {
    var logger = logger
    logger[metadataKey: "sub-system-req-id"] = "\(UUID())"
    logger.label = "com.my.sub.system"
}
```

or

```
func subSystemCall(logger: Logger) {
    var logger = Logger(label: "com.my.sub.system", delegating: Logger)
    logger[metadataKey: "sub-system-req-id"] = "\(UUID())"
}
```

which are kinda equivalent.

Due to what I consider a design mistake in SwiftLog, we're actually handing `label` to the logging backend _on construction_. So we can't actually just change it :(.

What we could do is when the label is being changed, re-construct the logging backend with the changed label.

Reply:
> we would like to pass metadata around

yes

> we would like to control the log-level (ie. pass it around)

yes, but why do we want to pass it around? I think Konrad suggested a good alternative - setting a level per subsystem, also, we might want to control the logging level in the configuration, passing loggers around will kill that

> we would like to pass the logging backend around

what do you mean by that? I think we definitely don't want to pass backends, backend is set up once, in the end application, all libraries should be absolutely backend-agnostic, imho (this is what actually got me a bit worried in this PR, basically as custom logging backend, even if its a noop), what am I missing here?

> we would like the sub-system to control the label

I personally like that, maybe because I'm used to that due to my java backed. But this is a "strong opinion, weakly held" :D, I just rolled with it in my projects since this was everything I could see in similar projects at the time. But if recommendation is different, I happy to change that opinion.

Reply:
TBH, my "ideal" solution would probably look like a logger per class (so two loggers, one in the `HTTPClient` and one in the `ConnectionPool`), with their respective labels, and all methods just accept `Logger.Metadata` or just a plain `requestId: String` or similar. But those I my crazy thoughts 😜 And after we have context, I would replace metadata/requestId with context. Not accepting anything in public API for now is also OK, util we have a context.

Reply:
Passing around `Logger.Metadata` is where I'm leaning for RediStack instead of a Logger instance, so it makes it easier for Logging backends to provide a mechanism for filtering logs by subsystem

Reply:
I wonder, can we maybe not expose logging in public API at all? That way we can do internally whatever we want, and then, after we have context and maybe honed in on a best pattern for logging, adopt it here?

Reply:
@artemredkin Passing logging between subsystems is useful for, for example, sampling.
`label`'s problem is that it's static - if you pass a single logger everywhere, label loses its meaning - it was supposed to identify a "subsystem", or more generally "pinpoint a place in the code" where your log is originated, but it doesn't do this any more. So maybe, "a place in the code where log was originated" in swift-log can only be a property of a method call (e.g. logger.info(message, label, metadata)), and not a `Logger` itself.

Reply:
Waait, does `source` field means that I have to set it in every method? something like:
```swift
func somefunc(logger: Logger) {
    logger.log("message", metadata: self.myMetdata, source: self.mySource)
}
```
or something? Thats a bit error prone, no? I think this might be better:
```swift
let logger: Logger
init() {
    self.logger = Logger("org.foo.bar", metadata: ["pool.limit": X])
}
func somefunc(context: Context) {
    self.logger.log("message", context: Context)
}
```
or at least a bit better? In my example I will never loose my label, metadata etc...

Reply:
@avolokhov I think we can sample at the log handler level, no? We don't have to update it in lots of places, just say `SamplingLogHandler(wrapping: RealLogHandler, sampling: ["label1", "label2", rate: 500])`, or even have a parser so we can configure it in the properties without recompiling anything

Reply:
Not if we want to sample a full request trace. We can workaround with setting a special field in metadata and then filter based on this field, but it's not very efficient and looks like a workaround.

Reply:
@artemredkin more along the lines of: We’d offer a new type with (tb bikeshed) name `LoggerWithSource` which is a struct containing Logger + source. So you’d do

   let logger =  LoggerWithSource(logger, “com.foo.bar”)
   logger.info(“hello”) // would log with source=com.foo.bar instead of MyModule

LoggerWithSource would have the same API as Logger. Obviously you’d never hand a LoggerWithSource to a subsys. If you called a subsys, you’d do `subSys(logger: loggerWithSource.logger)` to strip off the source again.

This way you should give you all the benefits of having automatic source as well as correctly passing all information around.

Reply:
Let's talk about the "source" things on the swift-log repo; it has some potential 👍 I'll have to try it out a bit to see if the new struct gets annoying or not too much, but it could work out.

Not to block this PR though, let's move there and PR can do what it wants for the time being I guess.

Reply:
That topic moved to: https://github.com/apple/swift-log/pull/135